### ISSUE TITLE
Using reduction variable causes misaligned write with CUDA target?

### ISSUE THREAD

**Issue by bluebear94** [AdaptiveCpp/AdaptiveCpp#527]
Minimal working example (adapted from page 9 of the [official cheat sheet] to make it compile):

~~~ c++
#include <algorithm>
#include <assert.h>

#include <CL/sycl.hpp>

using namespace cl::sycl;

int main() {
  queue myQueue;
  buffer<int> valuesBuf{1024};
  { // Initialize buffer on the host with 0, 1, 2, 3, ..., 1023
    host_accessor a{valuesBuf};
    std::iota(a.get_pointer(), a.get_pointer() + 1024, 0);
  }
  // Buffers with just 1 element to get the reduction results
  int sumResult = 0;
  buffer<int> sumBuf{&sumResult, 1};
  int maxResult = 0;
  buffer<int> maxBuf{&maxResult, 1};
  myQueue.submit(
      [&](handler &cgh) { // Input values to reductions are standard accessors
        auto inputValues = valuesBuf.get_access<access_mode::read>(
            cgh); // Create temporary objects describing variables with
        // reduction semantics
        accessor sumAccessor = sumBuf.get_access<access::mode::read_write>(cgh);
        accessor maxAccessor = maxBuf.get_access<access::mode::read_write>(cgh);
        auto sumReduction = reduction(sumAccessor, plus<int>());
        auto maxReduction = reduction(
            maxAccessor,
            maximum<int>()); // parallel_for performs two reduction operations
        // For each reduction variable, the implementation:
        // - Creates a corresponding reducer
        // - Passes a reference to the reducer to the lambda as a parameter
        cgh.parallel_for(range<1>{1024}, sumReduction, maxReduction,
                         [=](id<1> idx, auto &sum, auto &max) {
                           // plus<>() corresponds to += operator, so sum can be
                           // updated via += or combine()
                           sum += inputValues[idx];
                           // maximum<>() has no shorthand operator, so max
                           // can only be updated via combine()
                           max.combine(inputValues[idx]);
                         });
      });
  // sumBuf and maxBuf contain the reduction results once
  // the kernel completes
  assert(maxBuf.get_host_access()[0] == 1023 &&
         sumBuf.get_host_access()[0] == 523776);
}
~~~

When this is compiled (with `/opt/hipSYCL/bin/syclcc -Wall -Wextra -Wpedantic -O3 -std=c++17 -g --hipsycl-gpu-arch=sm_50 --cuda-path=/usr/local/cuda -L/usr/local/cuda/lib64 rtest.cpp -o rtest`) and run under `cuda-memcheck`, things don't look so good.

~~~
========= CUDA-MEMCHECK
[hipSYCL Warning] dag_direct_scheduler: Detected a requirement that is neither of discard access mode (SYCL 1.2.1) nor noinit property (SYCL 2020) that accesses uninitialized data. Consider changing to discard/noinit. Optimizing potential data transfers away.
[hipSYCL Warning] buffer_memory_requirement: Could not find embedded pointer in kernel blob for this requirement; do you have unnecessary accessors that are unused in your kernel?
[hipSYCL Warning] buffer_memory_requirement: Could not find embedded pointer in kernel blob for this requirement; do you have unnecessary accessors that are unused in your kernel?
========= Invalid __global__ write of size 4
=========     at 0x00000428 in /opt/hipSYCL/bin/../include/hipSYCL/glue/cuda/../generic/hiplike/hiplike_reducer.hpp:76:__hipsycl_kernel__ZTSZZZN7hipsycl4glue23hiplike_kernel_launcherILNS_2rt10backend_idE0ENS2_10cuda_queueEE4bindI24__hipsycl_unnamed_kernelLNS2_11kernel_typeE1ELi1EZZ4mainENKUlRNS_4sycl7handlerEE21_7clESB_EUlNS9_2idILi1EEERT_RT0_E35_26JNS9_6detail29accessor_reduction_descriptorINS9_8accessorIiLi1ELNS9_11access_modeE2ELNS9_6targetE0ELNS9_6access11placeholderE0EEENS9_4plusIiEEEENSL_ISR_NS9_7maximumIiEEEEEEEvNSD_IXT1_EEENS9_5rangeIXT1_EEES10_mT2_DpT3_ENKUlPNS2_8dag_nodeEE605_16clES15_ENKUlDpT_E646_41clIJNS0_16hiplike_dispatch28hiplike_reduction_descriptorISU_NS1B_15reduction_stageILi1EEEEENS1C_ISX_S1E_EEEEEDaS18_EUliDpRS17_E744_44
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x6a06ab1205f6b4d3 is misaligned
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/libcuda.so.1 (cuLaunchKernel + 0x2b8) [0x223718]
=========     Host Frame:/usr/local/cuda/lib64/libcudart.so.10.0 [0x1a23d]
=========     Host Frame:/usr/local/cuda/lib64/libcudart.so.10.0 [0x1a2c7]
=========     Host Frame:/usr/local/cuda/lib64/libcudart.so.10.0 (cudaLaunchKernel + 0x225) [0x4e3c5]
=========     Host Frame:./rtest [0xc17f]
=========     Host Frame:./rtest [0x1dced]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so (_ZN7hipsycl2rt10cuda_queue13submit_kernelERKNS0_16kernel_operationESt10shared_ptrINS0_8dag_nodeEE + 0x9e) [0x1252e]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x20f82]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt16kernel_operation8dispatchEPNS0_20operation_dispatcherESt10shared_ptrINS0_8dag_nodeEE + 0x56) [0x35886]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20multi_queue_executor15submit_directlyESt10shared_ptrINS0_8dag_nodeEEPNS0_9operationERKSt6vectorIS4_SaIS4_EE + 0x10b6) [0x200e6]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x2cca9]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20dag_direct_scheduler6submitESt10shared_ptrINS0_8dag_nodeEE + 0xd3c) [0x281fc]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x30094]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt13worker_thread4workEv + 0x244) [0x32b44]
=========     Host Frame:/lib/libstdc++.so.6 [0xcfbc4]
=========     Host Frame:/lib/libpthread.so.0 [0x9299]
=========     Host Frame:/lib/libc.so.6 (clone + 0x43) [0xff053]
=========
========= Program hit cudaErrorLaunchFailure (error 4) due to "unspecified launch failure" on CUDA API call to cudaMemcpyAsync. 
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/libcuda.so.1 [0x37b2c3]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so [0x76703]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so (_ZN7hipsycl2rt10cuda_queue13submit_memcpyERKNS0_16memcpy_operationESt10shared_ptrINS0_8dag_nodeEE + 0x2ad) [0x11c4d]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x210d2]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt16memcpy_operation8dispatchEPNS0_20operation_dispatcherESt10shared_ptrINS0_8dag_nodeEE + 0x57) [0x1b8f7]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20multi_queue_executor15submit_directlyESt10shared_ptrINS0_8dag_nodeEEPNS0_9operationERKSt6vectorIS4_SaIS4_EE + 0x10b6) [0x200e6]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x2cca9]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x2d53d]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x2ad63]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20dag_direct_scheduler6submitESt10shared_ptrINS0_8dag_nodeEE + 0xb19) [0x27fd9]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x30094]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt13worker_thread4workEv + 0x244) [0x32b44]
=========     Host Frame:/lib/libstdc++.so.6 [0xcfbc4]
=========     Host Frame:/lib/libpthread.so.0 [0x9299]
=========     Host Frame:/lib/libc.so.6 (clone + 0x43) [0xff053]
=========
[hipSYCL Error] from /home/kozet/gwaith/uofl-bioinformatics/gpgpu/hipSYCL/src/runtime/cuda/cuda_queue.cpp:221 @ submit_memcpy(): cuda_queue: Couldn't submit memcpy (error code = CUDA:4)
========= Program hit cudaErrorLaunchFailure (error 4) due to "unspecified launch failure" on CUDA API call to cudaEventQuery. 
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/libcuda.so.1 [0x37b2c3]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so [0x6c9de]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so (_ZNK7hipsycl2rt15cuda_node_event11is_completeEv + 0x2b) [0xf81b]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZNK7hipsycl2rt8dag_node11is_completeEv + 0x48) [0x22998]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt17dag_submitted_ops22update_with_submissionESt10shared_ptrINS0_8dag_nodeEE + 0x4dd) [0x30b5d]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt11dag_manager22register_submitted_opsESt10shared_ptrINS0_8dag_nodeEE + 0x4e) [0x2f90e]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20dag_direct_scheduler6submitESt10shared_ptrINS0_8dag_nodeEE + 0xe20) [0x282e0]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so [0x30094]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt13worker_thread4workEv + 0x244) [0x32b44]
=========     Host Frame:/lib/libstdc++.so.6 [0xcfbc4]
=========     Host Frame:/lib/libpthread.so.0 [0x9299]
=========     Host Frame:/lib/libc.so.6 (clone + 0x43) [0xff053]
=========
========= Program hit cudaErrorLaunchFailure (error 4) due to "unspecified launch failure" on CUDA API call to cudaEventSynchronize. 
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/libcuda.so.1 [0x37b2c3]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so [0x6c83e]
=========     Host Frame:/opt/hipSYCL/bin/../lib/hipSYCL/librt-backend-cuda.so (_ZN7hipsycl2rt15cuda_node_event4waitEv + 0x2b) [0xfd6b]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZN7hipsycl2rt20dag_multi_node_event4waitEv + 0x29) [0x24a69]
=========     Host Frame:/opt/hipSYCL/bin/../lib/libhipSYCL-rt.so (_ZNK7hipsycl2rt8dag_node4waitEv + 0x22) [0x23ee2]
=========     Host Frame:./rtest [0x1919b]
=========     Host Frame:./rtest [0x18862]
=========     Host Frame:./rtest [0xfec6]
=========     Host Frame:./rtest [0x699c]
=========     Host Frame:/lib/libc.so.6 (__libc_start_main + 0xd5) [0x27b25]
=========     Host Frame:./rtest [0x570e]
=========
[hipSYCL Error] from /home/kozet/gwaith/uofl-bioinformatics/gpgpu/hipSYCL/src/runtime/cuda/cuda_event.cpp:55 @ is_complete(): cuda_node_event: Couldn't query event status (error code = CUDA:4)
[hipSYCL Error] from /home/kozet/gwaith/uofl-bioinformatics/gpgpu/hipSYCL/src/runtime/cuda/cuda_event.cpp:66 @ wait(): cuda_node_event: cudaEventSynchronize() failed (error code = CUDA:4)
rtest: rtest.cpp:47: int main(): Assertion `maxBuf.get_host_access()[0] == 1023 && sumBuf.get_host_access()[0] == 523776' failed.
========= Error: process didn't terminate successfully
========= No CUDA-MEMCHECK results found
~~~

Is this a genuine bug with reduction variables or am I doing something wrong?

[official cheat sheet]: https://www.khronos.org/files/sycl/sycl-2020-reference-guide.pdf

**Comment by illuhad** [comment]
Thanks for reporting, it seems that #499 has introduced a regression that broke reductions with accessors. You can try reductions on top of USM pointers as a workaround. We will fix ASAP.

**Comment by illuhad** [comment]
Fix in PR #528 

This PR also adds a test case based on this code pattern where I have removed some things that are not necessary, such as the initialization of the buffers with a pointer.
